File: lambda/src/handler.ts
import {
  S3Client,
  PutObjectCommand,
  ListObjectsV2Command,
  GetObjectCommand,
} from "@aws-sdk/client-s3";
import type {
  APIGatewayProxyEvent,
  APIGatewayProxyResult,
} from "aws-lambda";

const BUCKET_NAME = process.env.BUCKET_NAME || "notes-data";

const isLocal = !!process.env.AWS_ENDPOINT_URL;

const s3 = new S3Client({
  region: process.env.AWS_REGION || "eu-central-1",
  ...(isLocal && {
    endpoint: process.env.AWS_ENDPOINT_URL,
    forcePathStyle: true,
    credentials: { accessKeyId: "test", secretAccessKey: "test" },
  }),
});

interface Note {
  id: string;
  title: string;
  content: string;
  createdAt: string;
}

const headers = {
  "Content-Type": "application/json",
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
  "Access-Control-Allow-Headers": "Content-Type",
};

async function createNote(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
  const body = JSON.parse(event.body || "{}");

  if (!body.title || !body.content) {
    return {
      statusCode: 400,
      headers,
      body: JSON.stringify({ error: "title and content are required" }),
    };
  }

  const note: Note = {
    id: `note-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    title: body.title,
    content: body.content,
    createdAt: new Date().toISOString(),
  };

  await s3.send(
    new PutObjectCommand({
      Bucket: BUCKET_NAME,
      Key: `${note.id}.json`,
      Body: JSON.stringify(note),
      ContentType: "application/json",
    })
  );

  return {
    statusCode: 201,
    headers,
    body: JSON.stringify(note),
  };
}

async function listNotes(): Promise<APIGatewayProxyResult> {
  const listResult = await s3.send(
    new ListObjectsV2Command({
      Bucket: BUCKET_NAME,
      Prefix: "note-",
    })
  );

  const notes: Note[] = [];

  for (const obj of listResult.Contents || []) {
    const getResult = await s3.send(
      new GetObjectCommand({
        Bucket: BUCKET_NAME,
        Key: obj.Key,
      })
    );
    const text = await getResult.Body?.transformToString();
    if (text) {
      notes.push(JSON.parse(text));
    }
  }

  notes.sort((a, b) => b.createdAt.localeCompare(a.createdAt));

  return {
    statusCode: 200,
    headers,
    body: JSON.stringify(notes),
  };
}

export async function handler(
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> {
  console.log("Event:", JSON.stringify(event, null, 2));

  if (event.httpMethod === "OPTIONS") {
    return { statusCode: 200, headers, body: "" };
  }

  try {
    switch (event.httpMethod) {
      case "POST":
        return await createNote(event);
      case "GET":
        return await listNotes();
      default:
        return {
          statusCode: 405,
          headers,
          body: JSON.stringify({ error: "Method not allowed" }),
        };
    }
  } catch (err) {
    console.error("Error:", err);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ error: "Internal server error" }),
    };
  }
}

File: lambda/handler.py
import json
import os
import time
import uuid

import boto3


BUCKET_NAME = os.environ.get("BUCKET_NAME", "notes-data")
AWS_REGION = os.environ.get("AWS_REGION", "eu-central-1")
AWS_ENDPOINT_URL = os.environ.get("AWS_ENDPOINT_URL")

s3_kwargs = {"region_name": AWS_REGION}
if AWS_ENDPOINT_URL:
    s3_kwargs["endpoint_url"] = AWS_ENDPOINT_URL
    s3_kwargs["aws_access_key_id"] = "test"
    s3_kwargs["aws_secret_access_key"] = "test"

s3 = boto3.client("s3", **s3_kwargs)

HEADERS = {
    "Content-Type": "application/json",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type",
}


def _response(status_code: int, body):
    return {
        "statusCode": status_code,
        "headers": HEADERS,
        "body": json.dumps(body) if not isinstance(body, str) else body,
    }


def _create_note(event):
    payload = json.loads(event.get("body") or "{}")
    title = payload.get("title")
    content = payload.get("content")

    if not title or not content:
        return _response(400, {"error": "title and content are required"})

    note = {
        "id": f"note-{int(time.time() * 1000)}-{uuid.uuid4().hex[:6]}",
        "title": title,
        "content": content,
        "createdAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }

    s3.put_object(
        Bucket=BUCKET_NAME,
        Key=f"{note['id']}.json",
        Body=json.dumps(note).encode("utf-8"),
        ContentType="application/json",
    )

    return _response(201, note)


def _list_notes():
    objects = s3.list_objects_v2(Bucket=BUCKET_NAME, Prefix="note-").get("Contents", [])
    notes = []

    for obj in objects:
        data = s3.get_object(Bucket=BUCKET_NAME, Key=obj["Key"])["Body"].read()
        notes.append(json.loads(data.decode("utf-8")))

    notes.sort(key=lambda n: n["createdAt"], reverse=True)
    return _response(200, notes)


def handler(event, context):
    method = event.get("httpMethod", "")

    if method == "OPTIONS":
        return {"statusCode": 200, "headers": HEADERS, "body": ""}

    try:
        if method == "POST":
            return _create_note(event)
        if method == "GET":
            return _list_notes()
        return _response(405, {"error": "Method not allowed"})
    except Exception as err:
        return _response(500, {"error": "Internal server error", "message": str(err)})
